crates/wasi/src/preview2/stream.rs

```diff
@@ -77,91 +77,78 @@ pub trait HostInputStream: Send + Sync {
     async fn ready(&mut self) -> Result<(), Error>;
 }
 
+#[derive(Debug)]
+pub enum OutputStreamError {
+    Closed,
+    LastOperationFailed(anyhow::Error),
+    Trap(anyhow::Error),
+}
+impl std::fmt::Display for OutputStreamError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            OutputStreamError::Closed => write!(f, "closed"),
+            OutputStreamError::LastOperationFailed(e) => write!(f, "last operation failed: {e}"),
+            OutputStreamError::Trap(e) => write!(f, "trap: {e}"),
+        }
+    }
+}
+impl std::error::Error for OutputStreamError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        match self {
+            OutputStreamError::Closed => None,
+            OutputStreamError::LastOperationFailed(e) | OutputStreamError::Trap(e) => {
+                Some(e.as_ref())
```

Comment:
Could this use `e.source()`? Otherwise errors will get double-printed because the error is printed in `Display` above and would additionally be printed when the `source` is printed here

Reply:
Thanks, I frequently get that one wrong